pilot/pkg/autoregistration/controller.go

```diff
@@ -702,8 +702,12 @@ func workloadEntryFromGroup(name string, proxy *model.Proxy, groupCfg *config.Co
 	if proxy.Metadata.Network != "" {
 		entry.Network = string(proxy.Metadata.Network)
 	}
-	if proxy.Locality != nil {
-		entry.Locality = util.LocalityToString(proxy.Locality)
+	// proxy.Locality can be unset when auto registration takes place, because
+	// its state is not fully initialized. Therefore, we check the bootstrap
+	// node.
+	if proxy.XdsNode != nil && proxy.XdsNode.Locality != nil {
+		entry.Locality = util.LocalityToString(proxy.XdsNode.Locality)
+		log.Infof("Setting Locality: %s for WLE: %s via XdsNode", entry.Locality, name)
 	}
```

Comment:
Not sure how this fix different with previous? 

As you can see from ads.go, we do set `proxy.Locality` in `setTopologyLabels`. So if no matter proxy labels or XdsNode contains locality info, we have used them to build proxy.Locality

Reply:
> Not sure how this fix different with previous?
Can you explain the previous bug? The PR didn't explain why `Proxy.Locality` was preferred over `Proxy.XdsNode`.

Thanks for the code pointer, I think I see the core problem:
In [DiscoveryServer.initializeProxy(con *Connection) error {}](https://github.com/istio/istio/blob/43a9c8b5a6cdbee3d46c3902cf7ab2b5b1fa2fea/pilot/pkg/xds/ads.go#L380), `s.WorkloadEntryController.OnConnect(con)` (which eventually calls `workloadEntryFromGroup`) is called before `s.computeProxyState(proxy, nil)` (which calls `setTopologyLabels(proxy)`)

Thus `proxy.Locality` is not populated until after this function, not before.

Reply:
I forgot about the issue reported very long ago. Need to recall

Reply:
Proxy.Locality is generated from labels or XdsNode. OF which the labels take priority, because in k8s, labels can be easy to control.

Reply:
Right, but see my comment above around the control flow in initializeProxy, setTopologyLabels happens too late.